lib/Sema/ConstraintSystem.cpp

```diff
@@ -63,16 +63,44 @@ void ConstraintSystem::solveConstraints()
     // TODO: Use Result to update the best solutions for each constraint.
 }
 
+bool ConstraintSystem::applyBind(Constraint *constraint, SystemState &state)
+{
+    auto *first = constraint->getFirstType();
+    auto *second = constraint->getSecondType();
+
+    if (first == second)
+        return true;
+
+    if (llvm::isa<glu::types::TypeVariableTy>(first)
+        && !llvm::isa<glu::types::TypeVariableTy>(second)) {
+        state.typeBindings[static_cast<glu::types::TypeVariableTy *>(first)]
+            = reinterpret_cast<glu::gil::Type *>(second);
```

Comment:
you can't reinterpret_cast, the types are not compatible. Marouan's PR changes so that they are types::TypeBase and not gil::Type, so you can rebase on his PR to fix that.